### DISCUSSION TITLE
Not quite sure what I'm missing here

### DISCUSSION THREAD

**Discussion by amritk** [gustavoguichard/string-ts#245] [Q&A] [answered]
Hello, this lib looks really cool but I'm not quite sure how to get it to work
![image](https://github.com/user-attachments/assets/8071c6a6-e9aa-438d-a6c3-033fbca0e29a)

```ts
  if (startsWith(props.example.type, 'http')) 
    console.log(props.example.type)
```
the type is not narrowed. I thought maybe because it is nested. But this doesn't work either

```ts
  const type = props.example.type  
  if (startsWith(type, 'http')) 
    console.log(type)
```

What am I missing here?

**Comment by gustavoguichard** [comment, marked as answer]
Hi @amritk , would you mind sending me a [TS playground example](https://www.typescriptlang.org/play) so I can better understand your problem and maybe fix a potential bug?

[edit]: If I undestood it right, it is not what you'd expect.
The resulting value of native `startsWith` is a `boolean` and in this case it will be `true`.
Check [this example](https://tsplay.dev/WY7XzN):
![Screenshot 2024-09-13 at 15 13 18](https://github.com/user-attachments/assets/d06783c4-a525-4560-82fd-a6ca5c662c98)

The goal of `string-ts` is to have a consistent result for the common string functions at both runtime and type levels, which is happening in this case.

Unfortunately it won't be enough for the TS compiler to understand the narrowing you expect here. Maybe in future iterations of TS ;)

**Reply by amritk** [reply]
Oh thats what it does, I thought it would narrow the type as well. Yea it doesn't work in typescript either [Playground](https://www.typescriptlang.org/play/?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBVDtUCkbDIUAAXlAAG9FCVktB4BJ8kyWSU0O0uQByWDQCgo1AAYSc3DFFAY0EEAtAAB9QAKCAxBIIJPAyNxFSqBVCFup5PrVQx2LEdRQAF51BqccXwaUCnkAX0UHsUCE4giOqHYiPpoAAFIINJBkFzqZHIABKekAPm5JQRofDNMgATZHICFKI9LpDKFIrFksg0u2svlAoTPsE8HcAXZzDDEeQ2fZEjjnsUQA)

**Reply by amritk** [reply]
I was trying to narrow the type by checking a nested property, actually its unrelated to this lib 😓

**Reply by gustavoguichard** [reply]
No worries, I'd expect it to work too and with every new version they've been improving the inference algorithm
